# Tasca base de dades NoSQL

Descripció:  
Comencem a familiaritzar-nos amb bases de dades NoSQL !!! Comencem amb uns quants exercicis bàsics

Objectius:  
- Utilitzar una base de dades MongoDB
- Interconnectar la base de dades MongoDB amb Python.

### Nivell 1
#### - Exercici 1
Crea una base de dades NoSQL utilitzant MongoDB. Afegeix-li algunes dades d'exemple que et permetin comprovar que ets capaç de processar-ne la informació de manera bàsica.

#### - Exercici 2
Connecta la base de dades NoSQL a Python utilitzant per exemple pymongo.

In [64]:
import pandas as pd
import pymongo
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

<ipython-input-64-6318dcb171c7>:6: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('display.max_colwidth', -1)


Ens conectem a mongo:

In [2]:
myclient = pymongo.MongoClient("mongodb://localhost:27017/")

Mirem quines bases de dades tenim a MongoDB:

In [3]:
print(myclient.list_database_names())

['Sprint18', 'acme', 'admin', 'config', 'local', 'mydatabase']


Creem una database i una col·lecció:

In [5]:
mydb = myclient["mydatabase"]

In [6]:
mycol = mydb["customers"]

In [9]:
# mycol.drop()

Provem d'afegir un sol registre, definim la id per defecte a 0:

In [10]:
mydict = {'_id': 0,
'name': 'Pepe',
'address': 'Gran Via 3'
         }

x = mycol.insert_one(mydict)

I diferents 'documents' a l'hora:

In [11]:
mylist = [
  { "_id": 1, "name": "John", "address": "Highway 37"},
  { "_id": 2, "name": "Peter", "address": "Lowstreet 27"},
  { "_id": 3, "name": "Amy", "address": "Apple st 652"},
  { "_id": 4, "name": "Hannah", "address": "Mountain 21"},
  { "_id": 5, "name": "Michael", "address": "Valley 345"},
  { "_id": 6, "name": "Sandy", "address": "Ocean blvd 2"},
  { "_id": 7, "name": "Betty", "address": "Green Grass 1"},
  { "_id": 8, "name": "Richard", "address": "Sky st 331"},
  { "_id": 9, "name": "Susan", "address": "One way 98"},
  { "_id": 10, "name": "Vicky", "address": "Yellow Garden 2"},
  { "_id": 11, "name": "Ben", "address": "Park Lane 38"},
  { "_id": 12, "name": "William", "address": "Central st 954"},
  { "_id": 13, "name": "Chuck", "address": "Main Road 989"},
  { "_id": 14, "name": "Viola", "address": "Sideway 1633"}
]
x = mycol.insert_many(mylist)

### Nivell 2
#### - Exercici 1
Carrega algunes consultes senzilles a un Pandas Dataframe. 

Provem de fer una consulta senzilla, agafem el primer document:

In [12]:
x = mycol.find_one()

print(x)

{'_id': 0, 'name': 'Pepe', 'address': 'Gran Via 3'}


Agafem tots els registres mitjançant una iteració:

In [13]:
for x in mycol.find():
    print(x)

{'_id': 0, 'name': 'Pepe', 'address': 'Gran Via 3'}
{'_id': 1, 'name': 'John', 'address': 'Highway 37'}
{'_id': 2, 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': 3, 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': 4, 'name': 'Hannah', 'address': 'Mountain 21'}
{'_id': 5, 'name': 'Michael', 'address': 'Valley 345'}
{'_id': 6, 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': 7, 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': 8, 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': 9, 'name': 'Susan', 'address': 'One way 98'}
{'_id': 10, 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': 11, 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': 12, 'name': 'William', 'address': 'Central st 954'}
{'_id': 13, 'name': 'Chuck', 'address': 'Main Road 989'}
{'_id': 14, 'name': 'Viola', 'address': 'Sideway 1633'}


In [14]:
for x in mycol.find({},{ "_id": 0, "name": 1, "address": 1 }):
    print(x)

{'name': 'Pepe', 'address': 'Gran Via 3'}
{'name': 'John', 'address': 'Highway 37'}
{'name': 'Peter', 'address': 'Lowstreet 27'}
{'name': 'Amy', 'address': 'Apple st 652'}
{'name': 'Hannah', 'address': 'Mountain 21'}
{'name': 'Michael', 'address': 'Valley 345'}
{'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'name': 'Betty', 'address': 'Green Grass 1'}
{'name': 'Richard', 'address': 'Sky st 331'}
{'name': 'Susan', 'address': 'One way 98'}
{'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'name': 'Ben', 'address': 'Park Lane 38'}
{'name': 'William', 'address': 'Central st 954'}
{'name': 'Chuck', 'address': 'Main Road 989'}
{'name': 'Viola', 'address': 'Sideway 1633'}


In [15]:
for x in mycol.find({},{ "_id": 0,"address": 0 }):
    print(x)

{'name': 'Pepe'}
{'name': 'John'}
{'name': 'Peter'}
{'name': 'Amy'}
{'name': 'Hannah'}
{'name': 'Michael'}
{'name': 'Sandy'}
{'name': 'Betty'}
{'name': 'Richard'}
{'name': 'Susan'}
{'name': 'Vicky'}
{'name': 'Ben'}
{'name': 'William'}
{'name': 'Chuck'}
{'name': 'Viola'}


Passem a fer queries senzilles i emmagatzemar-les en un df:

In [16]:
mydb = myclient["mydatabase"]
mycol = mydb["customers"]

df = pd.DataFrame(list(mycol.find()))

In [17]:
df

_id     name          address
0     0     Pepe       Gran Via 3
1     1     John       Highway 37
2     2    Peter     Lowstreet 27
3     3      Amy     Apple st 652
4     4   Hannah      Mountain 21
5     5  Michael       Valley 345
6     6    Sandy     Ocean blvd 2
7     7    Betty    Green Grass 1
8     8  Richard       Sky st 331
9     9    Susan       One way 98
10   10    Vicky  Yellow Garden 2
11   11      Ben     Park Lane 38
12   12  William   Central st 954
13   13    Chuck    Main Road 989
14   14    Viola     Sideway 1633

In [18]:
myquery = { "address": { "$regex": "^S" } }

mydoc = mycol.find(myquery)

df_list = []
for x in mydoc:
      df_list.append(x)
        
df1 = pd.DataFrame(df_list)
df1

_id     name       address
0    8  Richard    Sky st 331
1   14    Viola  Sideway 1633

In [19]:
myquery = { "address": "Park Lane 38" }

mydoc = mycol.find(myquery)

df_list = []
for x in mydoc:
      df_list.append(x)
        
df2 = pd.DataFrame(df_list)
df2

_id name       address
0   11  Ben  Park Lane 38

In [20]:
myquery = { "address": { "$gt": "S" } }

mydoc = mycol.find(myquery)

df_list = []
for x in mydoc:
      df_list.append(x)
        
df3 = pd.DataFrame(df_list)
df3

_id     name          address
0    5  Michael       Valley 345
1    8  Richard       Sky st 331
2   10    Vicky  Yellow Garden 2
3   14    Viola     Sideway 1633

### Nivell 3
#### - Exercici 1
Genera un resum estadístic de la informació que conté la base de dades.

Agafem una de les bases de dades que recomana MongoDB per treballar: Airbnb, amb les dades de Barcelona actualitzades el 12 d'Abril del 2021; per aconseguir dades més significatives per l'anàlisis estadístic. Pujarem les dades utilitzades ('listing.csv') al repositori corresponent: https://github.com/arampuig/Bases-de-dades-NoSQL. Ens conectem i accedim a la base de dades i la col·lecció (que hem creat prèviament mitjançant el MongoDB shell i afegint les dades d'Airbnb a una col·lecció) i la passem a pandas dataframe:

In [21]:
myclient = pymongo.MongoClient("mongodb://localhost:27017/")

print(myclient.list_database_names())

['Sprint18', 'acme', 'admin', 'config', 'local', 'mydatabase']


In [22]:
mydb = myclient["Sprint18"]

In [24]:
mycol = mydb["Airbnb"]

In [50]:
df = pd.DataFrame(list(mycol.find()))

In [51]:
df

_id        id  \
0      60cf64d344639f2ca04a35a9     18674   
1      60cf64d344639f2ca04a35aa     23197   
2      60cf64d344639f2ca04a35ab     32711   
3      60cf64d344639f2ca04a35ac     34241   
4      60cf64d344639f2ca04a35ad     34981   
...                         ...       ...   
18221  60cf64d444639f2ca04a7cd6  49155035   
18222  60cf64d444639f2ca04a7cd7  49155115   
18223  60cf64d444639f2ca04a7cd8  49155172   
18224  60cf64d444639f2ca04a7cd9  49155456   
18225  60cf64d444639f2ca04a7cda  49158347   

                                                    name   host_id  \
0        Huge flat for 8 people close to Sagrada Familia     71615   
1      Forum CCIB DeLuxe★Spacious &Elegant ★Large Bal...     90417   
2                       Sagrada Familia area - Còrsega 1    135703   
3                       Private Double room - Plaza Real     73163   
4                   VIDRE HOME PLAZA REAL on LAS RAMBLAS     73163   
...                                                  ...       ...   
18221                           Habitación 2 con Terraza  64805231   
18222                           Habitación 3 con Terraza  64805231   
18223                           Habitación 4 con Terraza  64805231   
18224                 Habitacion luninosa zona tranquila   8964574   
18225            BB- Bonito apartamento en casco antiguo  22076771   

              host_name neighbourhood_group  \
0      Mireia And Maria            Eixample   
1        Etain (Marnie)          Sant Martí   
2                  Nick              Gràcia   
3                Andres        Ciutat Vella   
4                Andres        Ciutat Vella   
...                 ...                 ...   
18221              Marc      Sants-Montjuïc   
18222              Marc      Sants-Montjuïc   
18223              Marc      Sants-Montjuïc   
18224           Lourdes          Nou Barris   
18225              Luis        Ciutat Vella   

                            neighbourhood  latitude longitude  \
0                      la Sagrada Família  41.40556   2.17262   
1                   el Besòs i el Maresme  41.41291   2.22063   
2      el Camp d'en Grassot i Gràcia Nova  41.40566   2.17015   
3                          el Barri Gòtic  41.38062   2.17517   
4                          el Barri Gòtic  41.37978   2.17623   
...                                   ...       ...       ...   
18221                               Sants  41.37949   2.13099   
18222                       Sants - Badal  41.37811   2.12925   
18223                               Sants  41.37998   2.13097   
18224                    Ciutat Meridiana  41.46193   2.17569   
18225                      el Barri Gòtic  41.38041   2.17913   

             room_type price minimum_nights number_of_reviews last_review  \
0      Entire home/apt   157              1                21  2019-11-01   
1      Entire home/apt   220              3                52  2019-12-15   
2      Entire home/apt   144              2                63  2020-02-22   
3         Private room    50             30                 8  2013-07-15   
4      Entire home/apt   243              4               149  2020-03-11   
...                ...   ...            ...               ...         ...   
18221     Private room    22             32                 0         NaN   
18222     Private room    22             32                 0         NaN   
18223     Private room    22             32                 0         NaN   
18224     Private room    18             30                 0         NaN   
18225  Entire home/apt    50             30                 0         NaN   

      reviews_per_month calculated_host_listings_count availability_365  
0                  0.22                             21               27  
1                  0.42                              2              359  
2                  0.53                              5                0  
3                  0.06                              2                0  
4    

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18226 entries, 0 to 18225
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   _id                             18226 non-null  object
 1   id                              18226 non-null  object
 2   name                            18215 non-null  object
 3   host_id                         18226 non-null  object
 4   host_name                       17687 non-null  object
 5   neighbourhood_group             18226 non-null  object
 6   neighbourhood                   18226 non-null  object
 7   latitude                        18226 non-null  object
 8   longitude                       18226 non-null  object
 9   room_type                       18226 non-null  object
 10  price                           18226 non-null  object
 11  minimum_nights                  18226 non-null  object
 12  number_of_reviews               18226 non-null

In [52]:
df.columns

Index(['_id', 'id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

Passem les columnes que contenen informació numèrica més rellevant a numeric type per descriure-la estadísticament:

In [53]:
 df[['price','minimum_nights', 'number_of_reviews','reviews_per_month', 
     'calculated_host_listings_count','availability_365']]= df[['price','minimum_nights', 'number_of_reviews',
                                                                'reviews_per_month', 'calculated_host_listings_count',
                                                                'availability_365']].apply(pd.to_numeric)

In [60]:
ind = df[df['price']==0].index

Eliminem els registres en que el preu per nit és igual a 0:

In [62]:
df.drop(ind, inplace=True)

In [63]:
df.describe()

price  minimum_nights  number_of_reviews  reviews_per_month  \
count  18221.000000    18221.000000       18221.000000       12560.000000   
mean     110.422260       12.293233          31.580265           0.916842   
std      429.124905       29.088617          63.903258           1.094043   
min        8.000000        1.000000           0.000000           0.010000   
25%       35.000000        1.000000           0.000000           0.140000   
50%       55.000000        3.000000           3.000000           0.480000   
75%      103.000000       30.000000          32.000000           1.340000   
max    11500.000000     1124.000000         808.000000          14.070000   

       calculated_host_listings_count  availability_365  
count                    18221.000000      18221.000000  
mean                        15.666868        175.125954  
std                         30.702464        141.157873  
min                          1.000000          0.000000  
25%                          1.000000          2.000000  
50%                          3.000000        179.000000  
75%                         14.000000        318.000000  
max                        164.000000        365.000000

CONCLUSIONS: 
- El preu mig per nit de l'oferta d'airbnb a Barcelona és de 110€, amb un mínim de 8€ i un màxim de 11500€. Tot i així observem que el 50è percentil es situa als 55€ i el 75è als 103€, donant-nos una idea més aproximada dels preus de la majoria d'habitacions.
- La mitjana de minim nits de reserva són 1 amb un màxim de 1124, i una mitjana de 12, tot i que el 50è percentil se situa a 3 nits.
- La mitjana de reviews per habitació són 31, amb un màxim de 808. El 50è percentil, però està a les 3 reviews u el 75è ja a les 32.
- La mitjana de reviews per mes és de 0.9, amb un màxim de 14.
- La mitjana de número d'inquilins acomulat per any és de 15.6, amb un mínim d'1 i màxim de 164. El 50è percentil està en 3 inquilins.
- La mitjana de disponibilitat de dies per any està a 175 de 365 dies a l'any.

Mirem els value_counts de les columnes categòriques més rellevants:

In [29]:
df['neighbourhood_group'].value_counts()

Eixample               6245
Ciutat Vella           4191
Sants-Montjuïc         2163
Sant Martí             1827
Gràcia                 1517
Sarrià-Sant Gervasi     768
Horta-Guinardó          592
Les Corts               380
Sant Andreu             286
Nou Barris              257
Name: neighbourhood_group, dtype: int64

El districte amb més nombre d'habitacions d'Airbnb és l'Eixample (6245), seguida per Ciutat Vella (4191). Els districtes minoritatis son Nou Barris (257), seguit de Sant Andreu (286) i Les Corts (380).

In [65]:
df['neighbourhood'].value_counts()

la Dreta de l'Eixample                          2036
el Raval                                        1412
el Barri Gòtic                                  1231
Sant Pere, Santa Caterina i la Ribera           1118
la Sagrada Família                              1039
l'Antiga Esquerra de l'Eixample                 984 
el Poble Sec                                    962 
la Vila de Gràcia                               951 
Sant Antoni                                     928 
la Nova Esquerra de l'Eixample                  744 
el Fort Pienc                                   512 
el Poblenou                                     429 
la Barceloneta                                  429 
Sants                                           409 
Sant Gervasi - Galvany                          392 
el Camp de l'Arpa del Clot                      329 
el Camp d'en Grassot i Gràcia Nova              325 
Hostafrancs                                     263 
les Corts                                     

In [66]:
df['neighbourhood'].value_counts().mean()

253.06944444444446

Entrant en barris més concretament veiem que els que tenen més habitacions són la Dreta de l'Eixample (2038), seguida del Raval (1412), el Barri Gòtic (1232), Sant Pere, Santa Caterina i la Ribera (1118) i la Sagrada Família (1039), barris clarament cèntrics i turístics.

La mitjana d'habitacions per barri és de 253, així podem observar quins estan clarament per sobre i per sota aquesta línea divisòria.

In [31]:
df['room_type'].value_counts()

Entire home/apt    9001
Private room       8710
Hotel room          283
Shared room         232
Name: room_type, dtype: int64

La gran majoria d'habitacions són o bé cases/apartaments sencers (9001) o bé habitacions privades (8710). Sent les habitacions d'hotel i les habitacions compartides molt més minoritàries (283 i 232, respectivament).